system-reinstall-bootc/src/lvm.rs

```diff
@@ -0,0 +1,103 @@
+use std::process::{Command, Stdio};
+
+use anyhow::Result;
+use bootc_mount::run_findmnt;
+use bootc_utils::CommandRunExt;
+use serde::Deserialize;
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct Lvs {
+    report: Vec<LvsReport>,
+}
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct LvsReport {
+    lv: Vec<LogicalVolume>,
+}
+
+#[derive(Debug, Deserialize, Clone)]
+pub(crate) struct LogicalVolume {
+    lv_name: String,
+    lv_size: String,
+    lv_path: String,
+    vg_name: String,
+}
+
+pub(crate) fn binary_available(binary: &str) -> bool {
+    Command::new("which")
+        .arg(binary)
+        .stdout(Stdio::null())
+        .stderr(Stdio::null())
+        .status()
+        .map(|status| status.success())
+        .unwrap_or(false)
+}
+
+pub(crate) fn parse_volumes(group: Option<String>) -> Result<Vec<LogicalVolume>> {
+    if !binary_available("lvs") {
+        tracing::debug!("lvs binary not found. Skipping logical volume check.");
+        return Ok(Vec::<LogicalVolume>::new());
+    }
+
+    let mut cmd = Command::new("lvs");
+    cmd.args([
+        "--reportformat=json",
+        "-o",
+        "lv_name,lv_size,lv_path,vg_name",
+    ]);
+
+    if let Some(group) = group {
+        cmd.arg(group);
+    }
+
+    let output: Lvs = cmd.run_and_parse_json()?;
+
+    Ok(output
+        .report
+        .iter()
+        .flat_map(|r| r.lv.iter().cloned())
+        .collect())
+}
+
+pub(crate) fn check_root_siblings() -> Result<Vec<String>> {
+    let all_volumes = parse_volumes(None)?;
+
+    // first look for a lv mounted to '/'
+    // then gather all the sibling lvs in the vg along with their mount points
+    let siblings: Vec<String> = all_volumes
+        .iter()
+        .filter(|lv| {
+            let mount = run_findmnt(&["-S", &lv.lv_path], None).unwrap_or_default();
+            if let Some(fs) = mount.filesystems.first() {
+                &fs.target == "/"
+            } else {
+                false
+            }
+        })
+        .flat_map(|root_lv| {
+            tracing::warn!("inside flat_map");
+            parse_volumes(Some(root_lv.vg_name.clone())).unwrap_or_default()
+        })
+        .filter_map(|lv| {
+            tracing::warn!("inside map");
+            let mount = run_findmnt(&["-S", &lv.lv_path], None).unwrap();
```

Comment:
Why not `?` instead of `unwrap()`?

Reply:
I'm not sure how to use `?` inside `filter_map` because it doesn't handle `Result`

Reply:
Ah right yeah, this is a general trap with combinators. Moving to explicit iteration is one option, but the other option is `try_fold()` which keeps in the combinator flow.

Reply:
Here's a little test example that shows the idea

```
#[test]
fn t() -> Result<()> {
    let v = ["1", "3", "", "4"];
    let r = v
        .into_iter()
        .try_fold(Vec::new(), |mut acc, r| -> anyhow::Result<_> {
            if r.is_empty() {
                Ok(acc)
            } else {
                acc.push(r.parse::<u32>()?);
                Ok(acc)
            }
        })?;
    dbg!(r);
    Ok(())
}
```

Try changing the array to have a non-integer value like `"oops"` in it to see the error path